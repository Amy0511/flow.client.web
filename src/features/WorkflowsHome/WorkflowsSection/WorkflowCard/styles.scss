@import "~@boomerang/boomerang-styles/lib/colors";

.c-workflow-card {
  display: flex;
  background-color: $white;
  flex-direction: column;
  justify-content: space-between;
  min-width: 22rem;
  width: 22rem;
  border: 0.125rem solid $teal-20;
  border-radius: 0.5rem;
  margin: 1rem;
  height: 9.5rem;

  .bx--overflow-menu {
    &:focus {
      outline: none;
      box-shadow: none;
    }
  }
  .bx--overflow-menu__icon {
    fill: $teal-20;
    &:hover {
      fill: $teal-20;
    }
  }
  .bx--overflow-menu-options__btn {
    color: $gray-50;
    text-decoration: none !important;
  }
  .bx--overflow-menu-options--open {
    border: solid 0.125rem $teal-20;
    border-radius: 0.5rem;
    z-index: 998;
  }
  .bx--overflow-menu-options:before {
    border-top-color: $teal-20;
    border-left-color: $teal-20;
  }

  .bx--overflow-menu-options__option {
    &:hover {
      background-color: $cool-gray-1;
    }
  }
}
.c-workflow-card__header {
  display: flex;
  justify-content: space-between;
  align-items: center;
}
.c-workflow-card__status {
  margin-left: 0.75rem;
  display: flex;
  align-items: center;
}
.b-workflow-card__status {
  color: $gray-50;
}
.b-workflow-card__circle {
  margin-right: 0.5rem;
  min-width: 0.75rem;
  max-width: 0.75rem;
  min-height: 0.75rem;
  max-height: 0.75rem;
  border-radius: 2rem;
  &.--published {
    background-color: $green-40;
  }
  &.--draft {
    background-color: $gray-20;
  }
}
.b-workflow-card__menu {
}
.c-workflow-card__info {
  display: flex;
  justify-content: space-between;
  height: 100%;
}
.c-workflow-card__icon {
  margin-left: 0.5rem;
  width: 5rem;
  display: flex;
  align-items: center;
  justify-content: flex-start;
}
.b-workflow-card__icon {
  width: 5rem;
  height: 4rem;
  margin-bottom: 1rem;
}
.c-workflow-card__description {
  display: flex;
  flex-direction: column;
  margin-top: 1.5rem;
  margin-left: 1rem;
  margin-right: 1rem;
  width: 100%;
  color: $gray-60;
}
.b-workflow-card__name {
  font-size: 1rem;
  font-weight: 600;
  margin-bottom: 0.5rem;
}
.b-workflow-card__description {
}

.b-workflow-card__launch {
  position: inherit;
  max-width: 70px;
  max-height: 70px;
}

.c-workflow-card__execute {
  max-width: 70px;
  max-height: 70px;
}
