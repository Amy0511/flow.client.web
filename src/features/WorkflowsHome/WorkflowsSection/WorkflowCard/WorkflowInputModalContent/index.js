import React, { Component } from "react";
import PropTypes from "prop-types";
import Body from "@boomerang/boomerang-components/lib/ModalContentBody";
import Footer from "@boomerang/boomerang-components/lib/ModalContentFooter";
import ConfirmButton from "@boomerang/boomerang-components/lib/ModalConfirmButton";
import TextArea from "@boomerang/boomerang-components/lib/TextArea";
import TextInput from "@boomerang/boomerang-components/lib/TextInput";
import SelectDropdown from "@boomerang/boomerang-components/lib/SelectDropdown";
import Toggle from "@boomerang/boomerang-components/lib/Toggle";
import INPUT_TYPES from "Constants/workflowInputTypes";
import "./styles.scss";

class WorkflowInputModalContent extends Component {
  static propTypes = {
    inputs: PropTypes.array.isRequired,
    executeWorkflow: PropTypes.func.isRequired
  };

  constructor(props) {
    super(props);

    const inputs = {};
    props.inputs.forEach(input => {
      inputs[input.key] = input.value;
    });

    this.state = {
      inputs,
      error: false
    };
  }

  componentDidMount() {
    this.validate();
  }

  handleBooleanChange = (e, key) => {
    const checked = e.target.checked;
    this.setState(prevState => ({ inputs: { ...prevState.inputs, [key]: checked } }), () => this.validate());
  };

  handleTextChange = (value, errors, key) => {
    this.setState(prevState => ({ inputs: { ...prevState.inputs, [key]: value } }), () => this.validate());
  };

  handleSelectChange = (option, key) => {
    this.setState(prevState => ({ inputs: { ...prevState.inputs, [key]: option.value } }), () => this.validate());
  };

  validate() {
    const errorInput = this.props.inputs.some(
      input =>
        input.required &&
        (this.state.inputs[input.key] === undefined ||
          this.state.inputs[input.key] === null ||
          (this.state.inputs[input.key] === "string" && this.state.inputs[input.key].length === 0))
    );
    this.setState({ error: errorInput });
  }

  renderInput = input => {
    const { key, type, defaultValue, label, required, validValues } = input;

    switch (type) {
      case INPUT_TYPES.BOOLEAN:
        return (
          <div className="b-workflow-inputs-modal-toggle">
            {required && <div className="s-workflow-inputs-modal-is-required">*</div>}
            <div className="b-workflow-inputs-modal-toggle__title">Value</div>
            <Toggle
              id={key}
              onChange={e => this.handleBooleanChange(e, key)}
              defaultChecked={defaultValue === "true" || false}
              theme="bmrg-white"
            />
          </div>
        );
      case INPUT_TYPES.SELECT:
        return (
          Array.isArray(validValues) && (
            <div className="b-workflow-inputs-modal-select">
              {required && <div className="b-workflow-inputs-modal-select__required">*</div>}
              <SelectDropdown
                onChange={option => this.handleSelectChange(option, key)}
                options={validValues.map(value => ({
                  label: value,
                  value: value
                }))}
                value={
                  this.state.inputs[key]
                    ? { label: this.state.inputs[key], value: this.state.inputs[key] }
                    : defaultValue
                }
                theme="bmrg-white"
                title={label}
                id={key}
                name={key}
              />
            </div>
          )
        );
      case INPUT_TYPES.TEXT_AREA:
        return (
          <div className="b-inputs-modal-text-area">
            <TextArea
              alwaysShowTitle
              title="Default Value"
              placeholder="Default Value"
              name={key}
              onChange={this.handleTextChange}
              value={defaultValue || ""}
              theme="bmrg-white"
              noValueText={`Enter a ${label}`}
              required={required}
            />
          </div>
        );
      default:
        return (
          <div className="b-workflow-inputs-modal-text-input">
            {required && <div className="s-workflow-inputs-modal-is-required">*</div>}
            <TextInput
              alwaysShowTitle
              title={label}
              placeholder={label}
              name={key}
              type={type}
              onChange={this.handleTextChange}
              value={defaultValue || ""}
              theme="bmrg-white"
              noValueText={`Enter a ${label}`}
              required={required}
            />
          </div>
        );
    }
  };

  render() {
    const { executeWorkflow, closeModal } = this.props;
    const { error } = this.state;

    return (
      <form>
        <Body className="b-workflow-inputs-modal-body">{this.props.inputs.map(this.renderInput)}</Body>
        <Footer className="b-workflow-inputs-modal-footer">
          <ConfirmButton
            type="submit"
            style={{ width: "40%" }}
            text={"EXECUTE"}
            disabled={error}
            onClick={e => {
              e.preventDefault();
              executeWorkflow({
                redirect: false,
                properties: this.state.inputs
              });
              closeModal();
            }}
            theme="bmrg-white"
          />
          <ConfirmButton
            type="submit"
            style={{ width: "40%" }}
            text={"EXECUTE AND VIEW"}
            disabled={error}
            onClick={e => {
              e.preventDefault();
              executeWorkflow({
                redirect: true,
                properties: this.state.inputs
              });
              closeModal();
            }}
            theme="bmrg-white"
          />
        </Footer>
      </form>
    );
  }
}

export default WorkflowInputModalContent;
