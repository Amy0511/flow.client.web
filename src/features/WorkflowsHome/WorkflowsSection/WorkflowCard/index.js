import React, { Component } from "react";
import axios from "axios";
import { notify, Notification } from "@boomerang/boomerang-components/lib/Notifications";
import PropTypes from "prop-types";
import { withRouter } from "react-router-dom";
import { OverflowMenu, OverflowMenuItem } from "carbon-components-react";
import { BASE_SERVICE_URL } from "Config/servicesConfig";
import Tooltip from "@boomerang/boomerang-components/lib/Tooltip";
import AlertModal from "@boomerang/boomerang-components/lib/AlertModal";
import ConfirmModal from "@boomerang/boomerang-components/lib/ConfirmModal";
import imgs from "Assets/icons";
import "./styles.scss";
import playButton from "./img/playButton.svg";

class WorkflowCard extends Component {
  static propTypes = {
    history: PropTypes.object.isRequired,
    workflow: PropTypes.object.isRequired,
    updateWorkflows: PropTypes.func.isRequired
  };

  handleExecute = () => {
    return axios
      .post(`${BASE_SERVICE_URL}/execute/${this.props.workflow.id}`)
      .then(response => {
        notify(<Notification type="success" title="Run Workflow" message="Succssfully ran workflow" />);
      })
      .catch(error => {
        notify(<Notification type="error" title="Something's wrong" message="Failed to run workflow" />);
      });
  };

  handleOnDelete = () => {
    // axios
    //   .delete(`${BASE_SERVICE_URL}/workflow/${this.props.workflow.id}`)
    //   .then(() => {
    //     notify(<Notification type="remove" title="SUCCESS" message="Workflow successfully deleted" />);
    //     this.props.updateWorkflows({workflowId:this.props.workflow.id, teamId:this.props.teamId});
    //     return;
    //   })
    //   .catch(() => {
    //     notify(<Notification type="error" title="SOMETHING'S WRONG" message="Your delete request has failed" />);
    //     return;
    //   });
    //this.props.updateWorkflows({ workflowId: this.props.workflow.id, teamId: this.props.teamId });
  };

  render() {
    const { workflow, history } = this.props;
    const menuOptions = [
      {
        itemText: "View Activity",
        onClick: () => history.push(`/activity/dfb6302b-62e0-4574-9062-727e4a37fc32`),
        primaryFocus: true
      },
      {
        itemText: "Edit Workflow",
        onClick: () => history.push(`/editor/${workflow.id}/designer`),
        primaryFocus: false
      },
      { itemText: "Delete", onClick: () => this.handleOnDelete(), primaryFocus: false }
    ];

    return (
      <div className="c-workflow-card">
        <div className="c-workflow-card__header">
          <div className="c-workflow-card__status">
            <div className={`b-workflow-card__circle --${workflow.status}`} />
            <label className="b-workflow-card__status">{workflow.status}</label>
          </div>
          <OverflowMenu className="b-workflow-card__menu" ariaLabel="" iconDescription="">
            {menuOptions.map(option => {
              return (
                <OverflowMenuItem
                  className="b-workflow-card__option"
                  requireTitle={false}
                  onClick={option.onClick}
                  itemText={option.itemText}
                  primaryFocus={option.primaryFocus}
                  key={option.itemText}
                />
              );
            })}
          </OverflowMenu>
        </div>
        <div className="c-workflow-card__info">
          <div className="c-workflow-card__icon">
            <img className="b-workflow-card__icon" src={imgs[workflow.icon]} alt="icon" />
          </div>
          <div className="c-workflow-card__description">
            <h2 className="b-workflow-card__name">{workflow.name}</h2>
            <p className="b-workflow-card__description">{workflow.description}</p>
            <span data-tip data-for={workflow.id} className="b-workflow-card-launch">
              <AlertModal
                ModalTrigger={() => (
                  <img src={playButton} className="b-workflow-card-launch__icon" alt="Execute workflow" />
                )}
                modalContent={closeModal => (
                  <ConfirmModal
                    title="Execute workflow?"
                    subTitleTop="It will run"
                    closeModal={closeModal}
                    affirmativeAction={this.handleExecute}
                    affirmativeText="Run"
                    theme="bmrg-white"
                  />
                )}
              />
            </span>
            <Tooltip id={workflow.id} place="bottom" theme="bmrg-white">
              Execute workflow
            </Tooltip>
          </div>
        </div>
      </div>
    );
  }
}

export default withRouter(WorkflowCard);
