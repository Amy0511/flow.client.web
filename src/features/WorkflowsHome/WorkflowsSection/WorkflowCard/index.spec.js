import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import { MemoryRouter } from "react-router";
import WorkflowCard from "./index";

jest.mock("./WorkflowInputModalContent", () => "WorkflowInputModalContent");

const mockfn = jest.fn();

const history = {};
const teamId = "1";
const updateWorkflows = mockfn;
const executeWorkflow = mockfn;
const deleteWorkflow = mockfn;
const workflow = {
  id: "789",
  name: "lucas-workflow-3",
  description: "blablabla",
  status: "draft",
  icon: "secure"
};

describe("WorkflowCard --- Snapshot", () => {
  it("Capturing Snapshot of WorkflowCard", () => {
    const renderedValue = renderer.create(
      <MemoryRouter>
        <WorkflowCard
          workflow={workflow}
          history={history}
          updateWorkflows={updateWorkflows}
          executeWorkflow={executeWorkflow}
          deleteWorkflow={deleteWorkflow}
          teamId={teamId}
        />
      </MemoryRouter>
    );
    expect(renderedValue).toMatchSnapshot();
  });
});

describe("WorkflowCard --- Shallow render", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <MemoryRouter>
        <WorkflowCard
          workflow={workflow}
          history={history}
          updateWorkflows={updateWorkflows}
          executeWorkflow={executeWorkflow}
          deleteWorkflow={deleteWorkflow}
          teamId={teamId}
        />
      </MemoryRouter>
    );
  });

  it("Render the DUMB component", () => {
    expect(wrapper.length).toEqual(1);
  });
});
