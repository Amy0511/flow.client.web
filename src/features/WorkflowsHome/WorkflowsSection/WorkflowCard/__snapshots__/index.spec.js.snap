// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkflowCard --- Snapshot Capturing Snapshot of WorkflowCard 1`] = `
<div
  className="c-workflow-card"
>
  <div
    aria-expanded={false}
    aria-haspopup={true}
    aria-label="card menu"
    className="b-workflow-card__menu bx--overflow-menu"
    onClick={[Function]}
    onClose={[Function]}
    onKeyDown={[Function]}
    open={false}
    role="button"
    style={
      Object {
        "position": "absolute",
        "right": "0",
      }
    }
    tabIndex={0}
  >
    <svg
      aria-label="overflow menu icon"
      className="bx--overflow-menu__icon"
      focusable="false"
      height={16}
      onClick={[Function]}
      onKeyDown={[Function]}
      preserveAspectRatio="xMidYMid meet"
      role="img"
      style={
        Object {
          "willChange": "transform",
        }
      }
      viewBox="0 0 16 16"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="8"
        cy="3"
        r="1"
      />
      <circle
        cx="8"
        cy="8"
        r="1"
      />
      <circle
        cx="8"
        cy="13"
        r="1"
      />
      <title>
        overflow menu icon
      </title>
    </svg>
  </div>
  <div
    className="c-workflow-card-info"
  >
    <div
      className="c-workflow-card__icon"
    >
      <img
        alt="icon"
        className="b-workflow-card__icon"
      />
    </div>
    <button
      className="c-workflow-card__description"
      onClick={[Function]}
    >
      <h2
        className="b-workflow-card__name"
      >
        lucas-workflow-3
      </h2>
      <p
        className="b-workflow-card__description"
      />
    </button>
  </div>
  <div
    className="b-workflow-card-launch"
  >
    <div
      onClick={[Function]}
      onKeyPress={[Function]}
      style={
        Object {
          "width": "fit-content",
        }
      }
    >
      <button
        className="bx--btn bx--btn--sm bx--btn--primary"
        disabled={false}
        tabIndex={0}
        type="button"
      >
        Execute Workflow
        <svg
          aria-hidden="true"
          aria-label="Run Workflow"
          className="bx--btn__icon"
          focusable="false"
          height={20}
          preserveAspectRatio="xMidYMid meet"
          role="img"
          style={
            Object {
              "willChange": "transform",
            }
          }
          viewBox="0 0 32 32"
          width={20}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21 16a6 6 0 1 1-6 6 6 6 0 0 1 6-6m0-2a8 8 0 1 0 8 8 8 8 0 0 0-8-8z"
          />
          <path
            d="M26 4H6a2 2 0 0 0-2 2v20a2 2 0 0 0 2 2h4v-2H6V12h22V6a2 2 0 0 0-2-2zM6 10V6h20v4z"
          />
          <path
            d="M19 19v6l5-3-5-3z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
